src/System.Numerics.Vectors/src/System/Numerics/Vector4_Intrinsics.cs

```diff
@@ -301,17 +301,24 @@ public static Vector4 SquareRoot(Vector4 value)
         /// <param name="value1">The source vector.</param>
         /// <param name="value2">The scalar value.</param>
         /// <returns>The result of the division.</returns>
-        [Intrinsic]
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         public static Vector4 operator /(Vector4 value1, float value2)
         {
-            float invDiv = 1.0f / value2;
```

Comment:
Why not just:
```
float invDiv = 1.0f / value2;
return value1 * invDiv
```

(I assume that `operator *(Vector4 left, Single right)` is intrinsic, or is it in the same boat)?

Also, why not just do `value1 / new Vector4(value2)` for both code-paths?

Reply:
> Why not just:

For the non `IsHardwareAccelerated` path?

> Also, why not just do `value1 / new Vector4(value2)` for both code-paths?

Is 4 divisions rather than 1 div + 4 mult for the non accelerated

Reply:
> (I assume that operator *(Vector4 left, Single right) is intrinsic, or is it in the same boat)?

It's not intrinsic but it's implemented correctly.

> Why not just:

Because it's slower.

Reply:
> For the non IsHardwareAccelerated path?

Yes.

> Is 4 divisions rather than 1 div + 4 mult for the non accelerated

Non-accelerated should only be on platforms which don't yet have the JIT support (we require SSE/SSE2 for x86 and ARM64 should also have Short Vector support). In my opinion it is not worth the increased maintainability costs just to squeeze out every bit of performance in the far edge cases.

@eerhardt may have other opinions...

Reply:
> we require SSE/SSE2 for x86 and ARM64 should also have Short Vector support

ARM32 doesn't have support for SIMD yet. Nor does .NET Framework x86 (though I don't know if this code is used to build System.Numerics.dll for .NET Framework).

Reply:
> In my opinion it is not worth the increased maintainability costs just to squeeze out every bit of performance in the far edge cases.

Which include versions of netfx and mono+unity. This function is used by `Vector4.Normalize()` which is a very heavily used path.

Reply:
> It's not intrinsic but it's implemented correctly.

@mikedn, its listed as `Intrinsic` (see [here](https://github.com/benaadams/corefx/blob/b4a14be9ac923f1b929179697a39e11a4374954e/src/System.Numerics.Vectors/src/System/Numerics/Vector4_Intrinsics.cs#L265)). So I was asking if it has the same issue as the division operator here (in which case we should also fix it up)

> Because it's slower.

Yes, doing a broadcast and divide is better than a divide, broadcast, and multiply. But the current code isn't doing that, so it may need a back-compat check.

Reply:
>  its listed as `Intrinsic`

Don't know if it actually is or not; couldn't find the code for `VectorN, scalar` so it might effect the other `VectorN` and other methods also.

`Intrinsic` is just a feel good lie in this code; don't think the Jit pays attention to it

Reply:
> Yes, doing a broadcast and divide is better than a divide, broadcast, and multiply. 

If broadcast was available, wouldn't want to get the reciprocal and multiply anyway as it is less precise. Just 4 divides in a row is terrible.

Reply:
> Intrinsic is just a feel good lie in this code; don't think the Jit pays attention to it

The presence of that attribute doesn't imply that the JIT treats the method as intrinsic.

Reply:
Wonder if there is a verification step that could be added to check if the annotation matches doing something special with the method...

Reply:
It seems to me that:
* We should be consistent in the handling of the accelerated and unaccelerated code paths
  * That is, if the accelerated code path does 4 divides (this includes via a single packed instruction), the unaccelerated code path needs to do the same, regardless of how fast or slow it is.
  * Having one do multiplication by reciprocal and the other do division will just lead to pain/confusion down the road when someone takes a dependency on the other behavior.
* If we are fixing this case, and there are other similar cases that could also easily be fixed, we should also do that from this PR

Reply:
* We should also ensure that this change is okay to make. It is a change in behavior (from doing multiplication by reciprocal to doing straight division) for one of the code paths. There may be existing code dependent on this behavior

Reply:
>  will just lead to pain/confusion down the road when someone takes a dependency on the other behavior.

At a guess its been a reciprocal since 2014; so any dependency is there; so using that argument the accelerated path should also the reciprocal at a cost of both performance and accuracy for compat in poor precision?

Reply:
/cc @jkotas any pointers which of the paths to take (bad numerics vs compat vs performance), or who can advise?

Reply:
Though I think if you taking a dependence on the exact value of a scalar floating point division you are in a bad place already; vectors are generally more consistently incorrect; whereas the incorrectness of scalars varies based on cpu?

Reply:
My opinion/preference is that:
* We take the compat change to do division instead of multiplication by reciprocal
* We ensure that IsHardwareAccelerated=true and IsHardwareAccelerated=false have the same behavior

Reply:
> vectors are generally more consistently incorrect; whereas the incorrectness of scalars varies based on cpu?

IEEE Floating-Point is deterministic, if nothing else. For a given machine, `divps` and four `divss` will return the same result for the same rounding-mode/configuration settings.

Reply:
System.Numerics.Vectors is an OOB NuGet package; and if you use it on current Unity, it will mean you get quite a performance hit for Vector4.Normalize vs current, as the Vectors were accelerated in [Mono 5.2.0 14 Aug 2017](http://www.mono-project.com/docs/about-mono/releases/5.2.0/) which Unity isn't at, yet.

Core and Framework may be fine (as they will take hardware accelerated path); but I don't think hitting the perf for games written in .NET on Unity is sensible?

Reply:
Unity does not generally pick up NuGet packages for things like this. They tend to have their own version that they can control.

Reply:
K, moved to single implementation and also changed Vector3 and Vector2